src/addr.c

```diff
@@ -174,29 +185,40 @@ void iface_init(char *ifname)
 				exit(1);
 			}
 
-			DBG("Found interface %s, address %s", ifa->ifa_name, buf);
+			DBG("Creating iface instance for interface %s, address %s", ifa->ifa_name, buf);
 			TAILQ_INSERT_TAIL(&iface_list, iface, link);
 
 			strlcpy(iface->ifname, ifa->ifa_name, sizeof(iface->ifname));
 			iface->ifindex = if_nametoindex(ifa->ifa_name);
 			iface->hostid = 1;
 			iface->sd = -1;
 		} else {
+			DBG("Found existing iface instance for interface name %s", ifa->ifa_name);
 			iface->unused = 0;
 		}
 
-		if (iface->inaddr.s_addr != ina.s_addr) {
-			if (is_zeronet(&iface->inaddr) || is_linklocal(&iface->inaddr)) {
-				iface->inaddr = ina;
-				iface->changed = 1;
+		if (ifa->ifa_addr->sa_family == AF_INET6) {
+			if (! IN6_ARE_ADDR_EQUAL(&iface->in6addr, &ina)) {
+				if (IN6_IS_ADDR_UNSPECIFIED(&iface->in6addr) || IN6_IS_ADDR_LINKLOCAL(&iface->in6addr)) {
```

Comment:
I am not 100% sure what the logic is for here. It looks to me like the code (in the original case for IPv4 below) is trying to find the "best" address, i.e. replacing a link-local one with a global one if available.

If this is the case then this code for IPv6 probably would need to take site-local addresses into account, too, and be more complicated so that a global one wouldn't be replaced by a site-local one.
Plus, with IPv6 link-local is a different beast than with IPv4. It is normal and usual for an interface to have both link-local and global addresses. But I guess in the first step, we could still declare a global address the "best" address to use.

Reply:
Yeah, not straight-forward at all. As long as we document the preferred order in the manual and as a comment close to the code, I'm happy with going for 1) something that works for most use-cases, 2) improve on it later.

Reply:
This is now implemented, hopefully in an understandable way.